lib/Dialect/ATen/ATenLoweringPass.cpp

```diff
@@ -0,0 +1,972 @@
+static MemRefType getShapeErasedMemRefType(MemRefType type) {
+  std::vector<int64_t> shape = type.getShape();
+  for(int i = 0; i < shape.size(); i++) {
+    shape[i] = -1;
+  }
+  return MemRefType::get(shape, type.getElementType(),
+                         type.getAffineMaps(), type.getMemorySpace());
+}
+
+/// Create a type cast to memref
+static Value MemRefTypeCast(PatternRewriter &builder, Value val) {
+  Type type = val.getType();
+
+  if (auto memrefTy = type.dyn_cast<MemRefType>()) {
+    MemRefType newType = getShapeErasedMemRefType(memrefTy);
+    return builder.create<MemRefCastOp>(val.getLoc(),
+                                          val, newType)
+      .getResult();
+  }
+  if (auto tensorTy = type.dyn_cast<TensorType>()) {
+      auto memRefType = mlir::MemRefType::get(tensorTy.getShape(),
+                                          tensorTy.getElementType(), {}, 0);
+      return typeCast(builder, val, memRefType);
+  }
+  return val;
+}
+
+static std::string getFullyMangledType(const Type ty) {
+  std::stringstream ret;
+
+  if (const MemRefType mrt = ty.dyn_cast<const MemRefType>()) {
+    ret << "M";
+    auto shape = mrt.getShape();
+    const Type elem = mrt.getElementType();
+    for (auto s : shape)
+      ret << s << "x";
+    ret << getFullyMangledType(elem);
+  } else if (FloatType ft = ty.dyn_cast<FloatType>()) {
+    ret << "F" << ft.getWidth();
+  } else if (const IntegerType it = ty.dyn_cast<const IntegerType>()) {
+    ret << "I" << it.getWidth();
+  } else if (const mlir::NPCOMP::aten::ATenListType alt =
+                 ty.dyn_cast<const mlir::NPCOMP::aten::ATenListType>()) {
+
+  } else {
+    Type t = ty;
+    t.dump();
+    assert(0 && "unhandled type in getFullyMangledType");
+  }
+  return ret.str();
+}
+
+// Mangle the argument shapes into the function name.  This is impractical for
+// a library-based implementation, since each different shape has to be
+// implemented by a different function.  A later pass will convert this to the
+// simple mangling scheme after we've converted to the LLVM dialect.
+static std::string getFullyMangledFuncName(std::string prefix,
+                                           FunctionType fnTy) {
+  std::string sep = "_";
+
+  auto resultTy = fnTy.getResults();
+  auto operTy = fnTy.getInputs();
+
+  std::string ret = prefix + "_AtenAcapOp_";
+  for (const Type t : resultTy)
+    ret = ret + sep + getFullyMangledType(t);
+  for (const Type t : operTy)
+    ret = ret + sep + getFullyMangledType(t);
+
+  return ret;
+}
+
+// Mangle the argument ranks into the function name.  This simpler version of
+// mangling allows us to implement most of the functions with only a few
+// variations. However, it means we need to convert from tensor types
+// with known size to tensor types with unknown size.
+static std::string getSimplyMangledType(const Type ty) {
+  std::stringstream ret;
+
+  if (const MemRefType mrt = ty.dyn_cast<const MemRefType>()) {
+    //    ret << "M";
```

Comment:
TODOs and cleaned up comments with explanation of what isn't supported yet?

Reply:
moved to issue.